controllers/effective_dnspolicies_reconciler.go

```diff
@@ -0,0 +1,268 @@
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	logger.V(1).Info("updating dns policies", "policies", len(policies))
+
+	for _, policy := range policies {
+		pLogger := logger.WithValues("policy", policy.GetLocator())
+
+		if policy.GetDeletionTimestamp() != nil {
+			pLogger.V(1).Info("policy marked for deletion, skipping")
+			continue
+		}
+
+		if accepted, _ := policyAcceptedFunc(policy); !accepted {
+			pLogger.V(1).Info("policy not accepted, skipping")
+			continue
+		}
+
+		listeners := r.listenersForPolicy(ctx, topology, policy, policyTypeFilterFunc)
+
+		if logger.V(1).Enabled() {
+			listenerLocators := lo.Map(listeners, func(item *machinery.Listener, _ int) string {
+				return item.GetLocator()
+			})
+			pLogger.V(1).Info("reconciling policy for gateway listeners", "listeners", listenerLocators)
+		}
+
+		var gatewayHasAttachedRoutes = false
+
+		clusterID, err := utils.GetClusterUID(ctx, r.client)
+		if err != nil {
+			return fmt.Errorf("failed to generate cluster ID: %w", err)
+		}
+
+		for _, listener := range listeners {
+			lLogger := pLogger.WithValues("listener", listener.GetLocator())
+
+			gateway := listener.Gateway
+			if listener.Hostname == nil || *listener.Hostname == "" {
+				lLogger.Info("skipping listener no hostname assigned")
+				continue
+			}
+
+			hasAttachedRoute := false
+			for _, statusListener := range gateway.Status.Listeners {
+				if string(listener.Name) == string(statusListener.Name) {
+					hasAttachedRoute = statusListener.AttachedRoutes > 0
+				}
+			}
+			if hasAttachedRoute {
+				gatewayHasAttachedRoutes = true
+			}
+
+			desiredRecord, err := desiredDNSRecord(gateway.Gateway, clusterID, policy, *listener.Listener)
+			if err != nil {
+				lLogger.Error(err, "failed to build desired dns record")
+				continue
+			}
+			if err = controllerutil.SetControllerReference(policy, desiredRecord, r.scheme); err != nil {
+				lLogger.Error(err, "failed to set owner reference on desired record")
+				continue
+			}
+			obj, err := controller.Destruct(desiredRecord)
+			if err != nil {
+				lLogger.Error(err, "unable to destruct dns record") // should never happen
+				continue
+			}
+			resource := r.client.Resource(DNSRecordResource).Namespace(desiredRecord.GetNamespace())
+
+			existingRecordObj, ok := lo.Find(topology.Objects().Children(listener), func(o machinery.Object) bool {
+				_, ok := o.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord)
+				return ok && o.GetNamespace() == listener.GetNamespace() && o.GetName() == dnsRecordName(listener.Gateway.Name, string(listener.Name))
+			})
+
+			//Update
+			if ok {
+				existingRecord := existingRecordObj.(*controller.RuntimeObject).Object.(*kuadrantdnsv1alpha1.DNSRecord)
+
+				//Deal with the potential deletion of a record first
+				if !hasAttachedRoute || len(desiredRecord.Spec.Endpoints) == 0 {
+					if logger.V(1).Enabled() {
```

Comment:
do we need this? I see it being used elsewhere without this check

Reply:
possibly not here since its not actually calculating anything additional to add into the log statements (probably was at one point and i removed it)